retrieval/targetmanager.go

```diff
@@ -14,139 +14,356 @@
+func (tm *TargetManager) removeTargets(f func(string) bool) {
+	if f == nil {
+		f = func(string) bool { return true }
+	}
+	var wg sync.WaitGroup
+	for src, targets := range tm.targets {
+		if !f(src) {
 			continue
 		}
+		wg.Add(len(targets))
+		for _, target := range targets {
+			go func(t Target) {
+				t.StopScraper()
+				wg.Done()
+			}(target)
+		}
+		delete(tm.targets, src)
+	}
+	wg.Wait()
+}
+
+// updateTargetGroup creates new targets for the group and replaces the old targets
+// for the source ID.
+func (tm *TargetManager) updateTargetGroup(tgroup *config.TargetGroup, cfg config.JobConfig) error {
+	newTargets, err := tm.targetsFromGroup(tgroup, cfg)
+	if err != nil {
+		return err
+	}
+	src := fullSource(cfg, tgroup.Source)
 
-		for _, targetGroup := range job.TargetGroup {
-			baseLabels := clientmodel.LabelSet{
-				clientmodel.JobLabel: clientmodel.LabelValue(job.GetName()),
+	tm.m.Lock()
+	defer tm.m.Unlock()
+
+	oldTargets, ok := tm.targets[src]
+	if ok {
+		// Replace the old targets with the new ones while keeping the state
+		// of intersecting targets.
+		for i, tnew := range newTargets {
+			var match Target
+			for i, told := range oldTargets {
+				if told == nil {
+					continue
+				}
+				if tnew.InstanceIdentifier() == told.InstanceIdentifier() {
+					match = told
+					oldTargets[i] = nil
+					break
+				}
 			}
-			for n, v := range m.globalLabels {
-				baseLabels[n] = v
+			// Update the exisiting target and discard the new equivalent.
+			// Otherwise start scraping the new target.
+			if match != nil {
+				match.Update(cfg, tgroup.Labels) // TODO(fabxc): wrong
+				newTargets[i] = match
+			} else {
+				go tnew.RunScraper(tm.sampleAppender, cfg.ScrapeInterval())
 			}
-			if targetGroup.Labels != nil {
-				for _, label := range targetGroup.Labels.Label {
-					baseLabels[clientmodel.LabelName(label.GetName())] = clientmodel.LabelValue(label.GetValue())
-				}
+		}
+		// Remove all old targets that disappeared.
+		for _, told := range oldTargets {
+			if told != nil {
+				go told.StopScraper()
 			}
+		}
+	} else {
+		// The source ID is new, start all target scrapers.
+		for _, tnew := range newTargets {
+			go tnew.RunScraper(tm.sampleAppender, cfg.ScrapeInterval())
+		}
+	}
+
+	if len(newTargets) > 0 {
+		tm.targets[src] = newTargets
+	} else {
+		delete(tm.targets, src)
+	}
+	return nil
+}
+
+// Pools returns the targets currently being scraped bucketed by their job name.
+func (tm *TargetManager) Pools() map[string][]Target {
+	pools := make(map[string][]Target)
+
+	for _, ts := range tm.targets {
+		for _, t := range ts {
+			job := string(t.BaseLabels()[clientmodel.JobLabel])
+			pools[job] = append(pools[job], t)
+		}
+	}
+	return pools
+}
+
+// UpdateConfig resets the managers target providers and job configurations as defined
+// by the new cfg. The state of targets that are valid in the new configuration remains unchanged.
+func (tm *TargetManager) UpdateConfig(cfg config.Config) error {
```

Comment:
:+1: for `ApplyConfig` - could be a potential interface for components that can handle config-reloads.